metals/src/main/scala/scala/meta/internal/tvp/MetalsTreeViewProvider.scala

```diff
@@ -316,6 +155,232 @@ class MetalsTreeViewProvider(
     }
   }
 
+  override def onCollapseDidChange(
+      params: TreeViewNodeCollapseDidChangeParams
+  ): Unit = getTrees().foreach(_.onCollapseDidChange(params))
+
+  override def parent(
+      params: TreeViewParentParams
+  ): TreeViewParentResult =
+    params.viewId match {
+      case Project =>
+        getTrees()
+          .map(_.parent(params.nodeUri))
+          .collectFirst { case Some(value) =>
+            value
+          }
+          .getOrElse(TreeViewParentResult(null))
+      case _ => TreeViewParentResult(null)
+    }
+
+  override def onVisibilityDidChange(
+      params: TreeViewVisibilityDidChangeParams
+  ): Unit = {
+    val trees = getTrees()
+    if (params.visible) {
+      params.viewId match {
+        case TreeViewProvider.Compile =>
+          ticks(params.viewId) = sh.scheduleAtFixedRate(
+            () => {
+              val all = trees.map(_.tickBuildTreeView()).collect {
+                case Some(value) => value
+              }
+              if (all.nonEmpty) {
+                languageClient.metalsTreeViewDidChange(
+                  TreeViewDidChangeParams(all.flatten.toArray)
+                )
+              }
+            },
+            1,
+            1,
+            TimeUnit.SECONDS,
+          )
+        case TreeViewProvider.Project =>
+          val toUpdate = trees.map(_.flushPendingProjectUpdates).collect {
+            case Some(value) => value
+          }
+          if (toUpdate.nonEmpty) {
+            languageClient.metalsTreeViewDidChange(
+              TreeViewDidChangeParams(toUpdate.flatten.toArray)
+            )
+          }
+        case _ =>
+      }
+    } else {
+      ticks.remove(params.viewId).foreach(_.cancel(false))
+    }
+  }
+
+  override def onBuildTargetDidCompile(
+      id: BuildTargetIdentifier
+  ): Unit = {
+    val toUpdate = getTrees().map(_.onBuildTargetDidCompile(id)).collect {
+      case Some(value) => value
+    }
+    if (toUpdate.nonEmpty) {
+      languageClient.metalsTreeViewDidChange(
+        TreeViewDidChangeParams(toUpdate.flatten.toArray)
+      )
+    }
+  }
+}
+
+class MetalsTreeFolderViewProvider(
+    folder: Folder,
+    buildTargets: BuildTargets,
+    compilations: () => TreeViewCompilations,
+    definitionIndex: GlobalSymbolIndex,
+    doCompile: BuildTargetIdentifier => Unit,
+    isBloop: () => Boolean,
+    statistics: StatisticsConfig,
+) {
+  private val classpath = new ClasspathSymbols(
+    isStatisticsEnabled = statistics.isTreeView
+  )
+  private val isVisible = TrieMap.empty[String, Boolean].withDefaultValue(false)
+  private val isCollapsed = TrieMap.empty[BuildTargetIdentifier, Boolean]
+  private val pendingProjectUpdates =
+    ConcurrentHashSet.empty[BuildTargetIdentifier]
+  val libraries = new ClasspathTreeView[AbsolutePath, AbsolutePath](
+    definitionIndex,
+    TreeViewProvider.Project,
+    s"libraries-${folder.uri.toString()}",
```

Comment:
Maybe we should have another folder level and libraries/project underneath? Possibly without that level if we only have one folder. We can change it in a later PR if this proves difficult.

Reply:
I added the screenshots, I think it's okay as a flattened structure. But I can change it, if you think it'll be better.